lib/OrdinaryDiffEqSDIRK/src/sdirk_caches.jl

```diff
@@ -654,21 +691,17 @@ end
 
 function alg_cache(
         alg::Union{Hairer4, Hairer42}, u, rate_prototype, ::Type{uEltypeNoUnits},
-        ::Type{uBottomEltypeNoUnits}, ::Type{tTypeNoUnits},
-        uprev, uprev2, f, t, dt, reltol, p, calck,
+        ::Type{uBottomEltypeNoUnits},
+        ::Type{tTypeNoUnits}, uprev, uprev2, f, t, dt, reltol, p, calck,
         ::Val{false}) where {uEltypeNoUnits, uBottomEltypeNoUnits, tTypeNoUnits}
-    if alg isa Hairer4
-        tab = Hairer4Tableau(constvalue(uBottomEltypeNoUnits), constvalue(tTypeNoUnits))
-    else
-        tab = Hairer42Tableau(constvalue(uBottomEltypeNoUnits), constvalue(tTypeNoUnits))
-    end
+    tab = get_sdirk_tableau(alg, constvalue(uBottomEltypeNoUnits), constvalue(tTypeNoUnits))
     γ, c = tab.γ, tab.γ
     nlsolver = build_nlsolver(alg, u, uprev, p, t, dt, f, rate_prototype, uEltypeNoUnits,
         uBottomEltypeNoUnits, tTypeNoUnits, γ, c, Val(false))
     Hairer4ConstantCache(nlsolver, tab)
 end
 
-@cache mutable struct Hairer4Cache{uType, rateType, uNoUnitsType, Tab, N} <:
+@cache mutable struct Hairer4Cache{uType, rateType, uNoUnitsType, Tab, N, StepLimiter} <:
```

Comment:
The point of the refactor is that these types don't need to exist. If there's a single SDIRKCache that just has different values for the tableau, then every SDIRK algorithm gives the same cache type, and so they compile all the same. If you have separate cache types, they will all compile different codes. So this doesn't meet the criteria. I recommend looking at the Rosenbrock methods, where a good chunk of them are all the same cache https://github.com/SciML/OrdinaryDiffEq.jl/blob/master/lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl#L748-L832, and all the same perform_step!, and it's just written based on the tableau size. This makes it so that compiling for one Rosenbrock compiles for all of them, effectively removing the per-algorithm compilation.